pkg/client/events/replay_client.go

```diff
@@ -212,128 +184,111 @@ func (rClient *replayClient[T]) Close() {
 	rClient.replayClientCancelCtx()
 }
 
-// goPublishEvents runs the work function returned by retryPublishEventsFactory,
-// re-invoking it according to the arguments to retry.OnError when the events bytes
-// observable returns an asynchronous error.
-func (rClient *replayClient[T]) goPublishEvents(ctx context.Context) {
-	// React to errors by getting a new events bytes observable, re-mapping it,
-	// and send it to replayObsCachePublishCh such that
-	// replayObsCache.Last(ctx, 1) will return it.
-	publishErr := retry.OnError(
-		ctx,
-		rClient.connRetryLimit,
-		eventsBytesRetryDelay,
-		eventsBytesRetryResetTimeout,
-		"goPublishEvents",
-		rClient.retryPublishEventsFactory(ctx),
-	)
-
-	// Since this function runs in a goroutine, we can't return the error to the
-	// caller. Instead, we panic.
-	if publishErr != nil {
-		panic(fmt.Errorf("EventsReplayClient[%T].goPublishEvents should never reach this spot: %w", *new(T), publishErr))
-	}
-}
-
-// retryPublishEventsFactory returns a function which is intended to be passed
-// to retry.OnError. The returned function pipes event bytes from the events
-// query client, maps them to typed events, and publishes them to the
-// replayObsCache replay observable.
-func (rClient *replayClient[T]) retryPublishEventsFactory(ctx context.Context) func() chan error {
-	return func() chan error {
-		eventsBzCtx, cancelEventsBzObs := context.WithCancel(ctx)
-		errCh := make(chan error, 1)
-
-		eventsBytesObs, err := rClient.eventsClient.EventsBytes(eventsBzCtx, rClient.queryString)
-		if err != nil {
-			cancelEventsBzObs()
-			errCh <- err
-			return errCh
+// goPublishEvents establishes and maintains an events subscription by
+// - Establishing EventsBytes subscription to events matching the query string
+// - Processes incoming events by decoding them from bytes to the generic type T
+// - Retries connection on failures up to connRetryLimit times with predefined delay
+// - Handles context cancellation to prevent resource leaks
+// - Cleans up connections when errors occur or context is canceled
+//
+// The method runs in a continuous loop until the context is canceled or retry limit is exceeded.
+func (rClient *replayClient[T]) goPublishEvents(ctx context.Context, publishCh chan<- T) {
+	numRetries := 0
```

Comment:
Does `numRetries` get reset after a successful retry? It seems like `numRetries` could accumulate between multiple intermittent errors otherwise, resulting in unexpected behavior.

Reply:
Good catch!